crates/flipperzero/src/furi/sync.rs

```diff
@@ -1,82 +1,108 @@
+impl FuriMutex {
+    const fn new() -> Self {
+        Self(AtomicPtr::new(ptr::null_mut()))
+    }
 
-        Mutex {
-            mutex,
-            data: UnsafeCell::new(data),
+    unsafe fn get(&self) -> *mut sys::FuriMutex {
+        let mutex = self.0.load(Ordering::Acquire);
+        if !mutex.is_null() {
+            mutex
+        } else {
+            self.create()
         }
     }
 
-    /// Acquires a mutex, blocking the current thread until it is able to do so.
-    pub fn lock(&self) -> furi::Result<MutexGuard<'_, T>> {
-        let status: Status =
-            unsafe { sys::furi_mutex_acquire(self.mutex.as_ptr(), u32::MAX).into() };
-        if status.is_err() {
-            return Err(status);
+    unsafe fn create(&self) -> *mut sys::FuriMutex {
+        let mutex = unsafe { sys::furi_mutex_alloc(MUTEX_TYPE) };
+
+        match self
+            .0
+            .compare_exchange(ptr::null_mut(), mutex, Ordering::Release, Ordering::Relaxed)
+        {
+            Ok(_) => mutex,
+            Err(global_ptr) => {
+                unsafe { sys::furi_mutex_free(mutex) };
+                global_ptr
+            }
         }
+    }
 
-        Ok(MutexGuard(self, PhantomData))
+    /// Attempts to acquire the mutex within `timeout` ticks, or polls without acquiring
+    /// if `timeout` is zero.
+    fn try_acquire(&self, timeout: u32) -> bool {
+        let status: Status = unsafe { sys::furi_mutex_acquire(self.get(), timeout).into() };
+        status.is_ok()
     }
 }
 
-unsafe impl<T: ?Sized + Send> Send for Mutex<T> {}
-unsafe impl<T: ?Sized + Send> Sync for Mutex<T> {}
-
-/// An RAII implementation of a "scoped lock" of a mutex.
-/// When this structure is dropped (falls out of scope), the lock will be unlocked.
-pub struct MutexGuard<'a, T: ?Sized + 'a>(&'a Mutex<T>, UnsendUnsync);
+impl Drop for FuriMutex {
+    fn drop(&mut self) {
+        let mutex = self.0.load(Ordering::Acquire);
+        if !mutex.is_null() {
+            unsafe { sys::furi_mutex_free(mutex) };
+        }
+    }
+}
 
-impl<T> Deref for MutexGuard<'_, T> {
-    type Target = T;
+unsafe impl RawMutex for FuriMutex {
+    const INIT: Self = FuriMutex::new();
+    type GuardMarker = GuardNoSend;
 
-    fn deref(&self) -> &Self::Target {
-        unsafe { &*self.0.data.get() }
+    fn lock(&self) {
+        // `INCLUDE_vTaskSuspend` is set to 1 in the Flipper Zero's FreeRTOS config, so as
+        // long as this timeout value matches `portMAX_DELAY` in the FreeRTOS config, this
+        // will block indefinitely as intended.
+        assert!(self.try_acquire(u32::MAX));
     }
-}
 
-impl<T> DerefMut for MutexGuard<'_, T> {
-    fn deref_mut(&mut self) -> &mut Self::Target {
-        unsafe { &mut *self.0.data.get() }
+    fn try_lock(&self) -> bool {
+        // Poll first so we don't block at all if we cannot acquire.
+        self.try_acquire(0) && self.try_acquire(1)
```

Comment:
Calling `xSemaphoreTake` with a timeout of 0 will acquire the lock if possible, so there's no need to call it a second time. The wording of the documentation is a bit ambiguous, but the behaviour is clarified in [this support thread](https://www.freertos.org/FreeRTOS_Support_Forum_Archive/September_2015/freertos_xSemaphoreTake_macro_with_delay_0_cannot_work_ec38d65fj.html).

Reply:
Heh, that is _not_ clear from the documentation. This is also how I originally assumed it would work, before reading said documentation and changing to the above approach 🤣 I'll change it back.